Automattic/woocommerce-payments
# Schedule individual subscription migrations and add manual migration tool under Status > Tools

Fixes #6525 

#### Changes proposed in this Pull Request









1. New Stripe Billing migration tool found under **WooCommerce > Status > Tools**
2. New scheduled action `wcpay_schedule_subscription_migrations` to fetch all of the Stripe Billing subscriptions and schedule individual actions to migrate them.

![image](https://github.com/Automattic/woocommerce-payments/assets/2275145/34ecc369-b474-4f6e-a264-1c50e26dfbe6)

For scalability, I decided to make our migrator class extend the `WCS_Background_Repairer` which we use in Subscriptions for large scale repair/upgrade scripts.

This class saves us writing similar code to handle scheduling the individual migrations in batches and handling timeouts etc all we need to do is just write a fetching function, see `get_items_to_repair()`.

Some other changes worth mentioning:

**Load WCPay Subscriptions service classes when migration feature is enabled**

Our migrator needs to access some of the WCPay Subscription classes (`includes/subscriptions/`) but we require the WCPay Subscriptions feature to be disabled before migrating. To load these classes when migrating, I've changed the logic to call `WC_Payments_Subscriptions::init()` when the Stripe Billing feature is enabled or the migration feature is enabled.

This should be low risk since we only create Stripe Billing subscriptions when the plugin is not active, and every other handling also checks if the subscription is a WCPay subscription before doing anything.

#### Testing instructions





While testing this PR, you'll need to activate the WC Subscriptions plugin and turn off the WCPay Subscriptions feature and vice versa. To help with this I've been using this snippet and commented out the lines I don't need:

```
// Creating WCPay Subscriptions to migrate
update_option( '_wcpay_feature_subscriptions', '1' );
deactivate_plugins( 'woocommerce-subscriptions/woocommerce-subscriptions.php' );

// Activate WC Subscriptions to test migrations
update_option( '_wcpay_feature_subscriptions', '0' );
activate_plugin( 'woocommerce-subscriptions/woocommerce-subscriptions.php' );
```

1. Enable the migrations feature by setting `_wcpay_feature_allow_subscription_migrations` to '1' in the wp-options table
2. Deactivate WC Subscriptions + enable WCPay Subscriptions
3. Purchase 1-5 WCPay Subscriptions
5. Activate WC Subscriptions + disable WCPay Subscriptions
6. Go to **WooCommerce > Status > Tools** and look for the new migrate tool
7. Click "Migrate Subscriptions" button
8. Confirm the button next to the tool is disabled to prevent running it multiple times
9. Open Action Scheduler and confirm there's a pending action `wcpay_schedule_subscription_migrations` that is scheduled to run within the next minute.
10. Wait for this to run on its own.
11. Once this is run, search Action Scheduler for pending actions with the hook `wcpay_migrate_subscription`.
12. There should be one pending `wcpay_migrate_subscription` action for each WCPay Subscription on your store.
13. Wait a minute to let these run and then go to **WooCommerce > Status > Logs** and search for `"migration"` and open the latest log.
14. Confirm the migrations have been successful.
15. Go back to the tools page and confirm the "Migrate Stripe Billing subscriptions" tool is no being added to the list of tools. Note: This tool should only load when there are subscriptions to migrate therefore after the migration is finished the tool should disappear.

-------------------


**Post merge**
